lektor/db.py

```diff
@@ -809,14 +813,126 @@ def thumbnail(self,
             width=width, height=height, mode=mode,
             upscale=upscale, quality=quality)
 
 
+def require_ffmpeg(f):
+    """Decorator to help with error messages for ffmpeg template functions."""
+    # If both ffmpeg and ffprobe executables are available we don't need to
+    # override the function
+    if locate_executable('ffmpeg') and locate_executable('ffprobe'):
+        return f
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        return Undefined('Unable to locate ffmpeg or ffprobe executable. Is '
+                         'it installed?')
+
+    return wrapper
+
+
+class Video(Attachment):
+    """Specific class for video attachments."""
+    def __init__(self, pad, data, page_num=None):
+        Attachment.__init__(self, pad, data, page_num)
+        self._video_info = None
+
+    def _get_video_info(self):
+        if self._video_info is None:
+            try:
+                self._video_info = get_video_info(self.attachment_filename)
+            except RuntimeError:
+                # A falsy value ensures we don't retry this video again
+                self._video_info = False
+        return self._video_info
+
+    @property
+    @require_ffmpeg
+    def width(self):
+        """Returns the width of the video if possible to determine."""
+        rv = self._get_video_info()
+        if rv:
+            return rv['width']
+        return Undefined('The width of the video could not be determined.')
+
+    @property
+    @require_ffmpeg
+    def height(self):
+        """Returns the height of the video if possible to determine."""
+        rv = self._get_video_info()
+        if rv:
+            return rv['height']
+        return Undefined('The height of the video could not be determined.')
+
+    @property
+    @require_ffmpeg
+    def duration(self):
+        """Returns the duration of the video if possible to determine."""
+        rv = self._get_video_info()
+        if rv:
+            return rv['duration']
+        return Undefined('The duration of the video could not be determined.')
+
+    @require_ffmpeg
+    def frame(self, seek=None):
+        """Returns a VideoFrame object that is thumbnailable like an Image."""
+        rv = self._get_video_info()
+        if not rv:
+            return Undefined('Unable to get video properties.')
+
+        if seek is None:
+            seek = rv["duration"] / 2
+        return VideoFrame(self, seek)
+
+
+class VideoFrame(object):
+    """Representation of a specific frame in a VideoAttachment.
+
+    This is currently only useful for thumbnails, but in the future it might
+    work like an ImageAttachment.
+    """
+
+    def __init__(self, video, seek):
+        self.video = video
+
+        if not isinstance(seek, timedelta):
+            seek = timedelta(seconds=seek)
+
+        if seek < timedelta(0):
+            raise ValueError("Seek distance must not be negative")
+        elif seek > video.duration:
```

Comment:
I think we want `elif video.duration and seek > video.duration:` here in case the video may be valid but we just couldn't determine duration. I don't know how likely that is, but we should try proceeding in that scenario.

Reply:
I guess we maybe could end up with a video without duration that has width and height. Will fix.

Reply:
Right. I have not idea if that would be really common or vanishingly rare. Thanks.